### PR TITLE
Add writing numpy tensors back to disk

### PR DESCRIPTION
**Pull request by seanmcl**
This is needed because while we will generally mutate the data in place, we will need to write new headers, e.g. for the potentially changed shape and row/column ordering. Moreover, for tensors we create from scratch, we will not in general have bits representing the header.

Tested by round-tripping in the python repl.

Unit tests are forthcoming.

### CODE REVIEW

File: TensorLib/NumpyRepr.lean [hunk 1]
```diff
@@ -514,5 +526,57 @@ def parseFile (path: System.FilePath) : IO (Except String NumpyRepr) := do
 
 end Parse
 
+section Save
+
+private def pushList (a : ByteArray) (xs : List UInt8) : ByteArray := xs.foldl (fun a b => a.push b) a
```
Comment (govereau) [inline TensorLib/NumpyRepr.lean:531]: Is it more efficient to use ByteArray.append, List.toArray ?

File: TensorLib/NumpyRepr.lean [hunk 2]
```diff
@@ -514,5 +526,57 @@ def parseFile (path: System.FilePath) : IO (Except String NumpyRepr) := do
...
+
+private def pushChars (a : ByteArray) (xs : List Char) : ByteArray := xs.foldl (fun a b => a.push (b.toUInt8)) a
+
+private def pushString (a : ByteArray) (xs : String) : ByteArray := pushChars a xs.toList
```
Comment (govereau) [inline TensorLib/NumpyRepr.lean:535]: Similar comment: String.toUTF8?

**Review by govereau — APPROVED** [review]

File: TensorLib/NumpyRepr.lean (hunk 1 above)
Comment (seanmcl) [reply TensorLib/NumpyRepr.lean:531]: Probably.  Will swap.

File: TensorLib/NumpyRepr.lean (hunk 2 above)
Comment (seanmcl) [reply TensorLib/NumpyRepr.lean:535]: ah didn't know about that one

**Pull request merged**